src/algorithms/cache.rs

```diff
@@ -0,0 +1,299 @@
+use crate::{Correctness, Guess, Guesser, DICTIONARY, MAX_MASK_ENUM};
+use once_cell::sync::OnceCell;
+use once_cell::unsync::OnceCell as UnSyncOnceCell;
+use std::borrow::Cow;
+use std::cell::Cell;
+use std::num::NonZeroU8;
+
+static INITIAL: OnceCell<Vec<(&'static str, f64, usize)>> = OnceCell::new();
+static PATTERNS: OnceCell<Vec<[Correctness; 5]>> = OnceCell::new();
+
+#[derive(Copy, Clone)]
+struct CacheValue(NonZeroU8);
+
+impl CacheValue {
+    fn new(val: u8) -> Self {
+        Self(NonZeroU8::new(val.wrapping_add(1)).unwrap())
+    }
+
+    fn get(&self) -> u8 {
+        self.0.get() - 1
+    }
+}
+
+const NUM_WORDS: usize = DICTIONARY.len();
+const CELL: Cell<Option<CacheValue>> = Cell::new(None);
+const ROW: [Cell<Option<CacheValue>>; NUM_WORDS] = [CELL; NUM_WORDS];
+thread_local! {
+    static COMPUTES: UnSyncOnceCell<Box<[[Cell<Option<CacheValue>>; NUM_WORDS]]>> = Default::default();
```

Comment:
I think we probably want to keep this an array of arrays rather than coerce the outer array into a slice. We know its size statically, so there's no point in erasing that information, and it may allow the compiler to optimize more.

Reply:
Sorry, I was writing up my comment. I am not sure of a stable API that allows us to create a box of an array without the array being on the stack. I think the unstable allocator api feature would help and I'm sure there is a way (possibly unsafe) to do it currently, but I am ignorant of that method.